packages/database/supabase/functions/create-space/index.ts

```diff
@@ -0,0 +1,220 @@
+        name: name.trim(),
+        url: url.trim().replace(/\/$/, ""),
+        platform,
+      },
+      {
+        onConflict: "url",
+        ignoreDuplicates: false,
+        count: "estimated",
+      },
+    )
+    .select()
+    .single();
+
+  if (result.error) return result;
+  const space_id = result.data.id;
+
+  // this is related but each step is idempotent, so con retry w/o transaction
+  const email = spaceAnonUserEmail(platform, result.data.id);
+  let anonymousUser: User | null = null;
+  {
+    const { error, data } = await supabase.auth.signInWithPassword({
+      email,
+      password,
+    });
+
+    if (
+      error &&
+      !(
+        error.code === "invalid_credentials" ||
+        error.message === "Invalid login credentials"
+      )
+    ) {
+      // Handle unexpected errors
+      return asPostgrestFailure(error.message, "authentication_error");
+    }
+    anonymousUser = data.user;
+  }
+  if (anonymousUser === null) {
+    const resultCreateAnonymousUser = await supabase.auth.admin.createUser({
+      email,
+      password,
+      email_confirm: true,
+    });
+    if (resultCreateAnonymousUser.error) {
+      return {
+        count: null,
+        status: resultCreateAnonymousUser.error.status || -1,
+        statusText: resultCreateAnonymousUser.error.message,
+        data: null,
+        error: new PostgrestError({
+          message: resultCreateAnonymousUser.error.message,
+          details:
+            typeof resultCreateAnonymousUser.error.cause === "string"
+              ? resultCreateAnonymousUser.error.cause
+              : "",
+          hint: "",
+          code: resultCreateAnonymousUser.error.code || "unknown",
+        }),
+      }; // space created but not its user, try again
+    }
+    anonymousUser = resultCreateAnonymousUser.data.user as User;
+  }
+  // NOTE: The next few steps could be done as the new user, except the SpaceAccess
+  const anonPlatformUserResult = await supabase
+    .from("PlatformAccount")
+    .upsert(
+      {
+        platform,
+        account_local_id: email,
+        name: `Anonymous of space ${space_id}`,
+        agent_type: "anonymous",
+        dg_account: anonymousUser.id,
+      },
+      {
+        onConflict: "account_local_id,platform",
+        ignoreDuplicates: false,
+        count: "estimated",
+      },
+    )
+    .select()
+    .single();
+  if (anonPlatformUserResult.error) return anonPlatformUserResult;
+
+  const resultAnonUserSpaceAccess = await supabase
+    .from("SpaceAccess")
+    .upsert(
+      {
+        space_id,
+        account_id: anonPlatformUserResult.data.id,
+        editor: true,
+      },
+      {
+        onConflict: "space_id,account_id",
+        ignoreDuplicates: false,
+        count: "estimated",
+      },
+    )
+    .select()
+    .single();
+  if (resultAnonUserSpaceAccess.error) return resultAnonUserSpaceAccess; // space created but not connected, try again
+  return result;
+};
+
+Deno.serve(async (req) => {
```

Comment:
Is there a way to make this type safe?
<img width="376" height="231" alt="image" src="https://github.com/user-attachments/assets/fbe1a1fa-e86a-4aa5-98fb-47b7ba866fa2" />